src/Microsoft.ML.Data/Transforms/ConcatTransform.cs

```diff
@@ -113,6 +127,218 @@ public sealed class TaggedArguments
+            {
+                // *** Binary format ***
+                // int: number of columns
+                // for each column:
+                //    columnInfo
+                int n = ctx.Reader.ReadInt32();
+                Contracts.CheckDecode(n > 0);
+                _columns = new ColumnInfo[n];
+                for (int i = 0; i < n; i++)
+                    _columns[i] = new ColumnInfo(ctx);
+            }
+            else
+                _columns = LoadLegacy(ctx);
+        }
+
+        private ColumnInfo[] LoadLegacy(ModelLoadContext ctx)
+        {
+            // *** Legacy binary format ***
+            // int: number of added columns
+            // for each added column
+            //   int: id of output column name
+            //   int: number of input column names
+            //   int[]: ids of input column names
+            // if version >= VersionAddedAliases
+            //   foreach column:
+            //      foreach non-null alias
+            //          int: index of the alias
+            //          int: string id of the alias
+            //      int: -1, marks the end of the list
+
+            int n = ctx.Reader.ReadInt32();
+            Contracts.CheckDecode(n > 0);
+            var names = new string[n];
+            var inputs = new string[n][];
+            for (int i = 0; i < n; i++)
+            {
+                names[i] = ctx.LoadNonEmptyString();
+                int numSources = ctx.Reader.ReadInt32();
+                Contracts.CheckDecode(numSources > 0);
+                inputs[i] = new string[numSources];
+                for (int j = 0; j < numSources; j++)
+                    inputs[i][j] = ctx.LoadNonEmptyString();
+            }
+
+            var aliases = new string[n][];
+            if (ctx.Header.ModelVerReadable >= VersionAddedAliases)
+            {
+                for (int i = 0; i < n; i++)
+                {
+                    var length = inputs[i].Length;
+                    aliases[i] = new string[length];
+                    if (ctx.Header.ModelVerReadable >= VersionAddedAliases)
+                    {
+                        for (; ; )
+                        {
+                            var j = ctx.Reader.ReadInt32();
+                            if (j == -1)
+                                break;
+                            Contracts.CheckDecode(0 <= j && j < length);
+                            Contracts.CheckDecode(aliases[i][j] == null);
+                            aliases[i][j] = ctx.LoadNonEmptyString();
+                        }
+                    }
+                }
+            }
+
+            var result = new ColumnInfo[n];
+            for (int i = 0; i < n; i++)
+                result[i] = new ColumnInfo(names[i],
+                    inputs[i].Zip(aliases[i], (name, alias) => (name, alias)));
+            return result;
+        }
+
         private sealed class Bindings : ManyToOneColumnBindingsBase
```

Comment:
>Bindings [](start = 29, length = 8)

You probably forgot to delete this peace of code.

Reply:
correct

---
In reply to: [217231847](https://github.com/dotnet/machinelearning/pull/896#discussion_r217231847) [](ancestors = 217231847)